File: iOS/Sample/NetworkViewController.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "NetworkViewController.h"

@interface NetworkViewController ()

@end

@implementation NetworkViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  self.navigationItem.title = @"Network";
}

- (IBAction)tappedGithubLitho:(UIButton*)sender {
  NSURL* url = [NSURL
      URLWithString:
          @"https://raw.githubusercontent.com/facebook/litho/main/docs/static/logo.png"];
  if (!url) {
    return;
  }

  [[[NSURLSession sharedSession] dataTaskWithURL:url
                               completionHandler:^(
                                   NSData* _Nullable data,
                                   NSURLResponse* _Nullable response,
                                   NSError* _Nullable error) {
                                 if (error && !data) {
                                   return;
                                 }
                                 NSLog(@"Got Image");
                               }] resume];
}

- (IBAction)tappedPOSTAPI:(UIButton*)sender {
  NSURL* url =
      [NSURL URLWithString:@"https://demo9512366.mockable.io/FlipperPost"];
  if (!url) {
    return;
  }

  NSMutableURLRequest* urlRequest = [NSMutableURLRequest requestWithURL:url];
  [urlRequest addValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
  [urlRequest addValue:@"application/json" forHTTPHeaderField:@"Accept"];
  NSDictionary* mapData =
      [[NSDictionary alloc] initWithObjectsAndKeys:@"Flipper",
                                                   @"app",
                                                   @"Its awesome",
                                                   @"remarks",
                                                   nil];
  NSError* error = nil;
  NSData* postData = [NSJSONSerialization dataWithJSONObject:mapData
                                                     options:0
                                                       error:&error];
  [urlRequest setHTTPBody:postData];
  [urlRequest setHTTPMethod:@"POST"];
  __weak NetworkViewController* weakSelf = self;
  [[[NSURLSession sharedSession]
      dataTaskWithRequest:urlRequest
        completionHandler:^(
            NSData* data,
            NSURLResponse* _Nullable response,
            NSError* _Nullable dataTaskError) {
          if (dataTaskError) {
            [weakSelf
                showAlertWithMessage:@"Received error in POST API response"];
            return;
          }
          if (data == nil) {
            [weakSelf
                showAlertWithMessage:@"No data received in POST API response"];
          } else {
            NSDictionary* dict =
                [NSJSONSerialization JSONObjectWithData:data
                                                options:0
                                                  error:&dataTaskError];
            NSLog(@"MSG-POST: %@", dict[@"msg"]);

            [weakSelf showAlertWithMessage:@"Received response from POST API"];
          }
        }] resume];
}

- (IBAction)tappedGetAPI:(UIButton*)sender {
  NSURL* url =
      [NSURL URLWithString:@"https://demo9512366.mockable.io/FlipperGet"];
  if (!url) {
    return;
  }
  __weak NetworkViewController* weakSelf = self;
  [[[NSURLSession sharedSession]
        dataTaskWithURL:url
      completionHandler:^(
          NSData* data,
          NSURLResponse* _Nullable response,
          NSError* _Nullable error) {
        if (error) {
          [weakSelf showAlertWithMessage:@"Received error in GET API response"];
          return;
        }
        if (data == nil) {
          [weakSelf
              showAlertWithMessage:@"No data received in GET API response"];
        } else {
          NSDictionary* dict = [NSJSONSerialization JSONObjectWithData:data
                                                               options:0
                                                                 error:&error];
          NSLog(@"MSG-GET: %@", dict[@"msg"]);
          [weakSelf showAlertWithMessage:@"Received response from GET API"];
        }
      }] resume];
}

- (void)showAlertWithMessage:(nonnull NSString*)msg {
  dispatch_async(dispatch_get_main_queue(), ^{
    UIAlertController* alertController = [self alertControllerForMessage:msg];
    [self presentViewController:alertController animated:true completion:nil];
  });
}

- (UIAlertController*)alertControllerForMessage:(nonnull NSString*)msg {
  UIAlertController* controller =
      [UIAlertController alertControllerWithTitle:@"Flipper"
                                          message:msg
                                   preferredStyle:UIAlertControllerStyleAlert];
  UIAlertAction* action =
      [UIAlertAction actionWithTitle:@"Ok"
                               style:UIAlertActionStyleDefault
                             handler:nil];
  [controller addAction:action];
  return controller;
}

@end

File: iOS/SampleSwift/SampleSwift/NetworkViewController.swift
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import UIKit

class NetworkViewController: UIViewController {

  @IBAction func tappedGetAPI(_ sender: UIButton) {
    let getURL = URL(string: "https://demo9512366.mockable.io/FlipperGet")!
    let dataTask = URLSession.shared.dataTask(with: getURL){ [weak self] (data, response, error) in
      guard let strongSelf = self else { return }
      guard let dataUnwrapped = data else {
        strongSelf.showAlert(message: "Received no data in GET API")
        return
      }

      if let errorUnwrapped = error {
        strongSelf.showAlert(message: "Received error in GET API Error:\(errorUnwrapped.localizedDescription)")
        return
      }

      let dict = try? JSONSerialization.jsonObject(with: dataUnwrapped, options: JSONSerialization.ReadingOptions.init(rawValue: 0)) as? [String: String]
      // As sonar cannot detect print() in Logs
      NSLog("MSG-GET: \(dict?["msg"] ?? "Did not find msg key in the received response")")
      strongSelf.showAlert(message: "Received response from GET API, please check the sonar network plugin for detailed response")
    }
    dataTask.resume()
  }

  @IBAction func tappedPOSTAPI(_ sender: UIButton) {
    guard let postURL = URL(string: "https://demo9512366.mockable.io/FlipperPost") else {
      showAlert(message: "Check the POST URL")
      return
    }
    var postRequest = URLRequest(url: postURL)
    postRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
    postRequest.addValue("application/json", forHTTPHeaderField: "Accept")

    let dict = ["app" : "Flipper", "remarks": "Its Awesome"]
    postRequest.httpBody = try! JSONSerialization.data(withJSONObject: dict, options: JSONSerialization.WritingOptions.init(rawValue: 0))
    postRequest.httpMethod = "POST"
    let dataTask = URLSession.shared.dataTask(with: postRequest){ [weak self] (data, response, error) in
      guard let strongSelf = self else { return }
      guard let dataUnwrapped = data else {
        strongSelf.showAlert(message: "Received no data in POST API")
        return
      }

      if let errorUnwrapped = error {
        strongSelf.showAlert(message: "Received error in POST API Error:\(errorUnwrapped.localizedDescription)")
        return
      }

      let dict = try? JSONSerialization.jsonObject(with: dataUnwrapped, options: JSONSerialization.ReadingOptions.init(rawValue: 0)) as? [String: String]
      // As sonar cannot detect print() in Logs
      NSLog("MSG-POST: \(dict?["msg"] ?? "Did not find msg key in the received response")")
      strongSelf.showAlert(message: "Received response from POST API, please check the sonar network plugin for detailed response")
    }
    dataTask.resume()
  }

  @IBAction func tappedFetchFBLitho(_ sender: UIButton) {
    let imageURL = URL(string: "https://raw.githubusercontent.com/facebook/litho/master/docs/static/logo.png")!
    let dataTask = URLSession.shared.dataTask(with: imageURL){ [weak self] (data, response, error) in
      guard let strongSelf = self else { return }
      guard let _ = data else {
        strongSelf.showAlert(message: "Received no data in Images API")
        return
      }

      if let errorUnwrapped = error {
        strongSelf.showAlert(message: "Received error in Images API Error:\(errorUnwrapped.localizedDescription)")
        return
      }

      // As sonar cannot detect print() in Logs
      NSLog("Got Image")
      strongSelf.showAlert(message: "Received Litho Image")
    }
    dataTask.resume()
  }

  func showAlert(message: String) {
    DispatchQueue.main.async {
      let alertController = UIAlertController.init(title: "Flipper", message: message, preferredStyle: .alert);
      let alertAction = UIAlertAction(title: "Okay", style: UIAlertAction.Style.default, handler: nil)
      alertController.addAction(alertAction)
      self.present(alertController, animated: true, completion: nil)
    }
  }
}
